flight/PiOS/Common/pios_px4flow.c

```
@@ -0,0 +1,286 @@
+/**
+ * @brief Updates the PX4Flow's orientation.
+ * @returns 0 for success or -1 for failure
+ */
+int32_t PIOS_PX4Flow_SetRotation(const struct Rotation rotation)
+{
+	if (PIOS_PX4Flow_Validate(dev) != 0)
+		return -1;
+
+	float rpy_R[3]  = {rotation.roll_D100/100.0f * DEG2RAD, rotation.pitch_D100/100.0f * DEG2RAD, rotation.yaw_D100/100.0f * DEG2RAD};
+
+	Euler2R(rpy_R, dev->Rsb);
+
+
+	return 0;
+}
+
+/**
+ * @brief Initialize the PX4Flow optical flow sensor
+ * \return none
+ * \param[in] PIOS_PX4Flow_ConfigTypeDef struct to be used to configure sensor.
+ *
+ */
+static int32_t PIOS_PX4Flow_Config(const struct pios_px4flow_cfg * cfg)
+{
+	// This function intentionally left blank.
+
+	return 0;
+}
+
+/**
+ * @brief Read current optical flow and sonar rangefinding measurements
+ * \param[out] optical_flow_data Struct containing translational x-, y-, and z-axis velocity
+ * \param[out] rangefinder_data Struct containing sonar ranging measurement along z-axis
+ * \return 0 for success or -1 for failure
+ */
+static int32_t PIOS_PX4Flow_ReadData(struct pios_sensor_optical_flow_data *optical_flow_data, struct pios_sensor_rangefinder_data *rangefinder_data)
+{
+	if (PIOS_PX4Flow_Validate(dev) != 0) {
+		return -1;
+	}
+
+	/* don't use PIOS_PX4Flow_Read and PIOS_PX4Flow_Write here because the task could be
+	 * switched out of context in between which would give the sensor less time to capture
+	 * the next sample.
+	 */
+	uint8_t addr_read[] = {
+		PIOS_PX4FLOW_FRAMECOUNTER_LSB
+	};
+
+	/* Optical flow structs */
+	struct I2C_Frame
+	{
+		uint16_t frame_count; // counts created I2C frames [#frames]
+		int16_t pixel_flow_x_sum_px10; // latest x flow measurement in pixels*10 [pixels]
+		int16_t pixel_flow_y_sum_px10; // latest y flow measurement in pixels*10 [pixels]
+		int16_t flow_comp_x_m1000; // x velocity*1000 [meters/sec]
+		int16_t flow_comp_y_m1000; // y velocity*1000 [meters/sec]
+		int16_t qual; // Optical flow solution quality [0: bad, 255: maximum quality]
```

Either way.  Right now we're ignoring the second byte by virtue of the assignment to uint8, but if we ever do any math/quality scaling/discarding of frames with poor quality we could be surprised.  So I think it makes sense to either change that to a uint8_t in the structure definition and then put a reserved byte after it, or to do if ((quality < 0) || (quality > UINT8_MAX)) { /* reject the frame or bound the quality */ }

There's arguments in favor of either approach, but silently ignoring the byte for now in this way leaves a trap for whoever hacks on it in the future.